internal/engine/procedures/clean/clean.go

```diff
@@ -0,0 +1,258 @@
+			c.cleanVar(&sfl.Variable)
+		},
+		StatementSQL: func(ss *parser.StatementSQL) {
+			c.cleanSQL(ss.Statement)
+		},
+		StatementReturn: func(sr *parser.StatementReturn) {
+			if sr.SQL != nil {
+				c.cleanSQL(sr.SQL)
+				if !returnable(sr.SQL) {
+					panic("RETURN statement must return a SELECT or have a RETURNING clause")
+				}
+			}
+		},
+		StatementProcedureCall: func(spc *parser.StatementProcedureCall) {
+			cleanedVars := make([]*string, len(spc.Variables))
+			for i, arg := range spc.Variables {
+				if arg == nil {
+					cleanedVars[i] = nil
+					continue
+				}
+				c.cleanVar(arg)
+				cleanedVars[i] = arg
+			}
+			spc.Variables = cleanedVars
+
+			_, ok := engine.Functions[strings.ToLower(spc.Call.Name)]
+			if !ok {
+				_, err := c.findProcedure(spc.Call.Name)
+				if err != nil {
+					panic(err)
+				}
+			}
+
+			spc.Call.Name = strings.ToLower(spc.Call.Name)
+		},
+		// loop targets
+		LoopTargetSQL: func(lts *parser.LoopTargetSQL) {
+			c.cleanSQL(lts.Statement)
+
+			if !returnable(lts.Statement) {
+				panic("LOOP target must be a SELECT or have a RETURNING clause")
+			}
+		},
+		// expressions
+		ExpressionMakeArray: func(ema *parser.ExpressionMakeArray) {
+			if len(ema.Values) == 0 {
+				panic("ARRAY must have at least one element")
+			}
+		},
+		ExpressionCall: func(ec *parser.ExpressionCall) {
+			_, ok := engine.Functions[strings.ToLower(ec.Name)]
+			if !ok {
+				_, err := c.findProcedure(ec.Name)
+				if err != nil {
+					panic(err)
+				}
+			}
+
+			ec.Name = strings.ToLower(ec.Name)
+		},
+		ExpressionVariable: func(ev *parser.ExpressionVariable) {
+			c.cleanVar(&ev.Name)
+		},
+		ExpressionComparison: func(ec *parser.ExpressionComparison) {
+			err := ec.Operator.Validate()
+			if err != nil {
+				panic(err)
+			}
+		},
+		ExpressionArithmetic: func(ea *parser.ExpressionArithmetic) {
+			err := ea.Operator.Validate()
+			if err != nil {
+				panic(err)
+			}
+		},
+	}
+
+	for _, stmt := range stmts {
+		stmt.Accept(&t)
+	}
+
+	return cleanedParams, c.cleanedSessionVars, nil
+}
+
+type cleaner struct {
+	currentSchema      *types.Schema
+	currentProc        *types.Procedure
+	sessionPrefix      string
+	knownSessionVars   map[string]*types.DataType
+	pgSchemaName       string
+	cleanedSessionVars map[string]*types.DataType
+}
+
+// cleanType ensures a type is fully qualified.
+// If the type is not a default type, it will be fully qualified.
+// If there is no schema specified, it will be assumed to be the current schema.
+// If there is a schema specified, it will use the alias.
+func (c *cleaner) cleanType(t *types.DataType) {
+	err := t.Clean()
+	if err != nil {
+		panic(err)
+	}
+}
+
+// cleanVar ensures a variable name is valid.
+func (c *cleaner) cleanVar(n *string) {
+	r := strings.ToLower(*n)
+
+	if len(r) == 0 || len(r) > 32 {
+		panic("variable names must be between 1 and 32 characters")
+	}
+
+	switch r[0] {
+	case '$':
+		// user-defined parameter
+		*n = "_param_" + r[1:]
```

Comment:
Yep, this is just to avoid collisions since we can't use `@` and`$` in plpgsql variables, and plpgsql is very pick with naming collisions (e.g. variables and columns cannot collide).